### ISSUE TITLE
Example doesn't work: use a loop to create exponential backoff for retries/until

### ISSUE THREAD

**Issue by jondoig** [ansible/ansible#75794]
### Summary

In the **Data manipulation** page of the User Guide, the section [Loops and list comprehensions](https://docs.ansible.com/ansible/latest/user_guide/complex_data_manipulation.html#loops-and-list-comprehensions) contains this example code:

```yaml
- name: retry ping 10 times with exponential backup delay
  ping:
  retries: 10
  delay: '{{item|int}}'
  loop: '{{ range(1, 10)|map('pow', 2) }}'
```

This code will not work:
- Ping ["does not make sense in playbooks"](https://docs.ansible.com/ansible/2.3/ping_module.html#synopsis)
- The loop expression contains unescaped quotes
- `retries` is used with `until` to repeat a task until a condition is met. Combining it with a loop creates a loop within a loop.

In any case, in this example it might be cleaner not to use `map` at all, but to use a simple integer sequence as the loop variable and apply the exponential function to the delay:
```yaml
  delay: "{{ item| power(2) | int }}"
  loop: "{{ range(1, 10) }}"
```

These changes can be made, e.g. using `uri` instead of `ping`, but it takes extra code to achieve the desired result. E.g. consider this playbook:

```yaml
---
- name: Test URI delay
  hosts: localhost
  tasks:
    - name: URI call
      vars:
        uri_task:
          status: -1
      uri:
        url: https://{{ site | default("google") }}.com
        timeout: 1
      failed_when: false
      retries: 1
      delay: "{{ item | pow(2) | int }}"
      until: uri_task.status != -1
      loop: "{{ range(1, 4) }}"
      register: uri_task
      when: uri_task.status != 200

    - name: Count URI calls
      set_fact:
        uri_count: "{{ uri_task.results | selectattr('skipped', 'undefined') | length }}"

    - name: Report result
      debug:
        msg: "{{ 'HTTP response ' ~ uri_task.results[uri_count | int - 1].status ~ ' on try ' ~ uri_count }}"
```

Running it with an invalid site name gives exponentially delayed retries, as desired:
`ansible-playbook test-uri.yml -e "site=googlez"`

```
TASK [URI call] *********************************************************************************************************************************************************
FAILED - RETRYING: URI call (1 retries left).
ok: [localhost] => (item=1)
FAILED - RETRYING: URI call (1 retries left).
ok: [localhost] => (item=2)
FAILED - RETRYING: URI call (1 retries left).
ok: [localhost] => (item=3)

TASK [Count URI calls] **************************************************************************************************************************************************
ok: [localhost]

TASK [Report result] ****************************************************************************************************************************************************
ok: [localhost] => {
    "msg": "HTTP response -1 on try 3"
}
```

Running it with the default valid site name ("google") skips the retries and skips remaining loop iterations:
`ansible-playbook test-uri.yml`

```
TASK [URI call] *********************************************************************************************************************************************************
ok: [localhost] => (item=1)
skipping: [localhost] => (item=2) 
skipping: [localhost] => (item=3) 

TASK [Count URI calls] **************************************************************************************************************************************************
ok: [localhost]

TASK [Report result] ****************************************************************************************************************************************************
ok: [localhost] => {
    "msg": "HTTP response 200 on try 1"
}
```

For the purpose of this documentation page, it might be better to choose a different example, e.g. converting each element in a list to Title Case:

```yaml
    - name: Create list
      set_fact:
        my_list:
          - foo
          - bar

    - name: Convert each list element to title case
      debug:
        msg: "{{ my_list | map('title') }}"
```

```
=> {
    "msg": [
        "Foo",
        "Bar"
    ]
}
```

### Issue Type

Documentation Report

### Component Name

 docs/docsite/rst/user_guide/complex_data_manipulation.rst

### Ansible Version

```console
$ ansible --version
ansible 2.10.6
  config file = /home/jdoig/repos/emmet-iac/ansible.cfg
  configured module search path = ['/home/jdoig/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /var/lib/awx/venv/sputnik-default/lib/python3.6/site-packages/ansible
  executable location = /var/lib/awx/venv/sputnik-default/bin/ansible
  python version = 3.6.12 (default, Sep 15 2020, 12:49:50) [GCC 4.8.5 20150623 (Red Hat 4.8.5-37)]
```

### Configuration

```console
$ ansible-config dump --only-changed
```

### OS / Environment

Red Hat Enterprise Linux Server release 7.9 (Maipo)

### Additional Information

The suggested change replaces a broken code example with a working one.

The broken example code was introduced as described in [this comment on issue 20226](https://github.com/ansible/ansible/issues/20226#issuecomment-726853897) which should perhaps be reopened.

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

**Comment by ansibot** [comment]
Files identified in the description:
* [`docs/docsite/rst/user_guide/complex_data_manipulation.rst`](https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/user_guide/complex_data_manipulation.rst)

If these files are incorrect, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/devel/ISSUE_HELP.md)

**Comment by bcoca** [comment]
> Ping "does not make sense in playbooks"

Not true, its great in detecting 'non boostrapped' hosts that are not managed by ansible yet, so i use it as a condition to call my bootstrapping plays (basically install the python that is needed).


>    The loop expression contains unescaped quotes
True, the outer ones were meant to be double quotes

>    retries is used with until to repeat a task until a condition is met. Combining it with a loop creates a loop within a loop.

Also true, not a problem 'per se' but  I'll be sticking to a single loop in the update

**Comment by jondoig** [comment]
Thanks @bcoca.

To clarify, my comment about ping is a quote from the linked Ansible module doco, so maybe that page could do with an update?

I agree it's useful to document an exponential backoff example, and thanks for doing so. Your code led me to implement something similar as a helper role.

Perhaps this example fits better somewhere focused on URI rather than on the Data Manipulation page? Maybe a Redhat blog post? Or an addition to [Anthony Critelli's post](https://www.redhat.com/sysadmin/ansible-web-endpoints).

**Comment by bcoca** [comment]
i believe the comment was more about the 'pong' option than about the ping module

it is in data manipulation as an example of creating a range with a specific distribution, i'll probably move to uri as most people will want that for accessing an API and avoiding getting throttled (wait_for_conneciton makes my initial example mostly obsolete).

**Comment by mveroone** [comment]
Hi @bcoca 
Sorry for digging up an old subject but I stumbled upon this very issue today.
And while I found the initial solution quite elegant, it indeed doesn't work unless I forcing the task to never fail, which we obviously don't want to.

Are you aware of any plans to introduce a better exponential back-off solution ?

In the meantime, I'll try to hack something with block/rescue, counting the failed/succeeded iterations and really failing if all iterations failed.

**Comment by bcoca** [comment]
the original example was not about a loop, but a wait/sleep task, loop delay is not settable per iteration, i was looking for alternative but have not found one yet. I might just revert to what it should have been before it was 'improved' to be more 'ansible like'

**Comment by mveroone** [comment]
For the record, here is how I implemented my exponential backoff
While not ideal, it works as expected.
The only downside is that it doesn't stop the loop when result becomes successful. But that's acceptable

```
- block:
  - name: Wait with exponential backoff
    uri:
      url: "http://example.org/api/status/cyz"
      method: GET 
      return_content: yes 
    register: uri_status
    retries: 1
    until: <success condition based on `uri_status` register>
    delay: '{{ item|int }}' 
    loop: "{{ range(2, 16)|map('pow', 2)|list }}" 
  rescue:
  - fail:
      msg: Timed out 
    when: (uri_status.results|length - uri_status.results|selectattr("failed")|list|length) == 0
```

**Comment by bcoca** [comment]
For 'breaking out' of the loop, something like this:
```
     when: not (uri_status|default({})).get('results', [{}])[-1] is success
```
basically 'once you succeed, skip rest'

**Comment by mveroone** [comment]
Amazing. Everything's possible with Ansible, that's why I love it so much.
Thanks

**Comment by mveroone** [comment]
Actually, I thought it worked, but after further testing, it seems the "results" item is only populatied once the full task is finished and the register only contains the result of the current iteration when queries in the "when" section.

**Comment by bcoca** [comment]
`results` should have 'all previous' iterations, not the current (since that is the one being run), also note that the first iteration has an 'undefined' `results`

**Comment by mveroone** [comment]
So this *might* be because my ansible version is outdated(we haven't made the leap past ansible 2.9 yet :/ )but not here :
```test.yml
- hosts: localhost
  tasks:
  - name: test
    debug:
      var: reg
    loop: "{{ range(1,5)|list }}"
    when: reg is undefined or reg.item|int < 3
    register: reg
  - debug:
      var: reg
``` 
gives
```
TASK [test] **********************************************************************************************************
ok: [localhost] => (item=1) => {
    "ansible_loop_var": "item",
    "item": 1,
    "reg": "VARIABLE IS NOT DEFINED!"
}
ok: [localhost] => (item=2) => {
    "ansible_loop_var": "item",
    "item": 2,
    "reg": {
        "ansible_loop_var": "item",
        "changed": false,
        "failed": false,
        "item": 1,
        "reg": "VARIABLE IS NOT DEFINED!"
    }
}
ok: [localhost] => (item=3) => {
    "ansible_loop_var": "item",
    "item": 3,
    "reg": {
        "ansible_loop_var": "item",
        "changed": false,
        "failed": false,
        "item": 2,
        "reg": {
            "ansible_loop_var": "item",
            "changed": false,
            "failed": false,
            "item": 1,
            "reg": "VARIABLE IS NOT DEFINED!"
        }
    }
}
skipping: [localhost] => (item=4) 

TASK [debug] *********************************************************************************************************
ok: [localhost] => {
    "reg": {
        "changed": false,
        "msg": "All items completed",
        "results": [
            {
                "ansible_loop_var": "item",
                "changed": false,
                "failed": false,
                "item": 1,
                "reg": "VARIABLE IS NOT DEFINED!"
            },
            {
                "ansible_loop_var": "item",
                "changed": false,
                "failed": false,
                "item": 2,
                "reg": {
                    "ansible_loop_var": "item",
                    "changed": false,
                    "failed": false,
                    "item": 1,
                    "reg": "VARIABLE IS NOT DEFINED!"
                }
            },
            {
                "ansible_loop_var": "item",
                "changed": false,
                "failed": false,
                "item": 3,
                "reg": {
                    "ansible_loop_var": "item",
                    "changed": false,
                    "failed": false,
                    "item": 2,
                    "reg": {
                        "ansible_loop_var": "item",
                        "changed": false,
                        "failed": false,
                        "item": 1,
                        "reg": "VARIABLE IS NOT DEFINED!"
                    }
                }
            },
            {
                "ansible_loop_var": "item",
                "changed": false,
                "item": 4,
                "skip_reason": "Conditional result was False",
                "skipped": true
            }
        ]
    }
}
```
No "results" until task is finished

**Comment by jborean93** [comment]
I've fixed the invalid quotes in the example with https://github.com/ansible/ansible/pull/79356. I agree with @bcoca and that we don't want to complicate an example with more in depth options. If you still have further issues with the docs please open another issue with more details. If you have general questions please send them towards the mailing list or IRC https://docs.ansible.com/ansible/latest/community/communication.html.
